src/timer/timer_game_calendar.h

```diff
@@ -109,6 +109,70 @@ class TimerGameCalendar {
 	static Month month; ///< Current month (0..11).
 	static Date date; ///< Current date in days (day counter).
 	static DateFract date_fract; ///< Fractional part of the day.
+
+
+	static constexpr int DAYS_IN_YEAR = 365; ///< days per year
+	static constexpr int DAYS_IN_LEAP_YEAR = 366; ///< sometimes, you need one day more...
+	static constexpr int MONTHS_IN_YEAR = 12; ///< months per year
+
+	static constexpr int SECONDS_PER_DAY = 2;   ///< approximate seconds per day, not for precise calculations
+
+	/*
+	 * ORIGINAL_BASE_YEAR, ORIGINAL_MAX_YEAR and CALENDAR_DAYS_TILL_ORIGINAL_BASE_YEAR are
+	 * primarily used for loading newgrf and savegame data and returning some
+	 * newgrf (callback) functions that were in the original (TTD) inherited
+	 * format, where 'TimerGameCalendar::date == 0' meant that it was 1920-01-01.
+	 */
+
+	/** The minimum starting year/base year of the original TTD */
+	static constexpr Year ORIGINAL_BASE_YEAR = 1920;
+	/** The original ending year */
+	static constexpr Year ORIGINAL_END_YEAR = 2051;
+	/** The maximum year of the original TTD */
+	static constexpr Year ORIGINAL_MAX_YEAR = 2090;
+
+	/**
+	 * Calculate the date of the first day of a given year.
+	 * @param year the year to get the first day of.
+	 * @return the date.
+	 */
+	static constexpr Date DateAtStartOfYear(Year year)
+	{
+		int32_t year_as_int = static_cast<int32_t>(year);
+		uint number_of_leap_years = (year == 0) ? 0 : ((year_as_int - 1) / 4 - (year_as_int - 1) / 100 + (year_as_int - 1) / 400 + 1);
+
+		return (DAYS_IN_YEAR * year_as_int) + number_of_leap_years;
+	}
+
+	/**
+	 * Calculate the year of a given date.
+	 * @param date The date to consider.
+	 * @return the year.
+	 */
+	static Year DateToYear(Date date);
+
+	/** The absolute minimum & maximum years in OTTD */
+	static constexpr Year MIN_YEAR = 0;
+
+	/** The default starting year */
+	static constexpr Year DEF_START_YEAR = 1950;
+	/** The default scoring end year */
+	static constexpr Year DEF_END_YEAR = ORIGINAL_END_YEAR - 1;
+
+	/**
+	 * MAX_YEAR, nicely rounded value of the number of years that can
+	 * be encoded in a single 32 bits date, about 2^31 / 366 years.
+	 */
+	static constexpr Year MAX_YEAR = 5000000;
+
+	static constexpr Year INVALID_YEAR = -1; ///< Representation of an invalid year
+	static constexpr Date INVALID_DATE = -1; ///< Representation of an invalid date
 };
 
+/** The date of the first day of the original base year. */
+static constexpr TimerGameCalendar::Date CALENDAR_DAYS_TILL_ORIGINAL_BASE_YEAR = TimerGameCalendar::DateAtStartOfYear(TimerGameCalendar::ORIGINAL_BASE_YEAR);
+
+/** The date of the last day of the max year. */
+static constexpr TimerGameCalendar::Date MAX_CALENDAR_DATE = TimerGameCalendar::DateAtStartOfYear(TimerGameCalendar::MAX_YEAR + 1) - 1;
```

Comment:
This unbalance of what variable is scoped inside `TimerGameCalendar` and which is not, will drive us absolutely mental. You have to check the source every time to know which is where. Simplicity of code is important for future-us; surprises like this really makes the difference between "this is easy to work with" and "wtf happened here" :D :)

So I would suggest either of the two (but there might be other suggestions):

(1) Move all constants outside of the class, and all use the `CALENDAR_` prefix.

This would ofc mean that later overloading and moving to a base-class will be more difficult. But you stepped into that one already with having 2 defined outside; you will need to redefine them anyway. So that is just a result of your current choice :D

(2) Create a new class `Calendar`, where you put all the constants in. So that you get `Calendar::INVALID_YEAR`, and `Calendar::MAX_DATE`.

It is fine to leave the functions in `TimerGameCalendar`, as that way the second class is just a namespace. So on that note, it doesn't even have to be a class or struct, but it can also just be a namespace definition. Just a namespace you can't inherit, and as it was your goal to make it easier to split economy, a class (or struct) might be more to the point.

Reply:
I went with option 2, a new CalendarTime class (named to avoid confusion with economy.cpp when we have EconomyTime later).

I also switched to precomputed magic numbers for `DAYS_TILL_ORIGINAL_BASE_YEAR` and `MAX_DATE`, because defining those outside the class was worse, IMO.

Ticks also got their own class, for consistency and since I had to change all their references anyway.